Beam eigenstrain classes and eigenstrain transfer

sveerara opened the pull request:
Adds beam eigenstrain base class, thermal eigenstrain class and transfer of eigenstrains from vectorpostprocessor and CSV files. 

refs #10313

sveerara committed:
Eigenstrain classes and eigenstrain transfer #10313

sveerara commented:
@bwspenc, @aeslaughter: Can you please help me figure out the problem with the documentation? I am getting the following error:
`MooseDocs.tree.app_syntax: Failed to execute the MOOSE executable '/opt/civet/build_0/moose/modules/combined/combined-opt':
list index out of range`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/179976/) on 3b3d608 : invalidated by @bwspenc

bwspenc commented:
@sveerara I'm not sure what's going on with the documentation build.  I just invalidated it to try again.

sveerara commented:
Thanks @bwspenc. The other two tests that are failing are also related to documentation I think. MooseDocs/test/tree.syntax test is failing in Pthreads and yaml test is failing in Marmot.

sveerara commented:
@acasagran and @bwspenc, please review this PR when you have time.

bwspenc commented on modules/combined/test/tests/beam_eigenstrain_transfer/master.i:
```diff
@@ -0,0 +1,224 @@
+  [./elasticity]
+    type = ComputeElasticityBeam
+    youngs_modulus = 2.60072400269
+    poissons_ratio = -0.9998699638
```
The constants for youngs_modulus and poissons_ratio are kind of odd. Where do they come from?

sveerara replied:
The values of E, A and I and length of beam are based on the timoshenko beam example from Prathap and Bhasyam (1982) and the example is in beam/static/timoshenko_small_strain_y.i

In beam/static/euler_small_strain_y.i, I converted the timoshenko beam to a Euler-Bernoulli beam by keeping the geometry of the beam constant but by varying the shear modulus. Basically if alpha (= kAGL^2/EI) is a large number, the beam would behave like a Euler-Bernoulli beam irrespective of whether the beam is slender or thick. For alpha = 2e6, the shear modulus has to be 1e4 which results in a poisson's ratio of -0.9998. This is within the allowable range for poisson's ratio.

bwspenc replied:
If you could put in a short comment to that effect in this input file, that would be helpful. Since this is a test of the eigenstrain transfer and not of the behavior of the beam, it also might be worth considering using a Timoshenko beam in this example.

bwspenc commented on modules/tensor_mechanics/doc/content/documentation/systems/Materials/ComputeEigenstrainBeamFromVariable.md:
```diff
@@ -0,0 +1,17 @@
+
+## Description
+
+`ComputeBeamEigenstainFromVariable` class takes in a set of 3 displacement or rotation variables or both and uses them to calculate the displacement and rotational beam eigenstrain in the global coordinate system. If either displacement or rotational aux variables are provided, they should be provided in all three global coordinate directions.
```
Either drop 'class' or add 'The' to the beginning of the sentence.

bwspenc commented on modules/tensor_mechanics/doc/content/documentation/systems/Transfers/VectorPostprocessorTransfer.md:
```diff
@@ -0,0 +1,17 @@
+
+## Description
+
+`VectorPostprocessorTransfer` transfers vectorpostprocessor data from the sub applications to auxvariables in the master application. The VectorPostprocessor data in the `subapp_component` direction is extracted along with the data corresponding to the list of variable vectors. This data is linearly interpolated and mapped to auxvariables in the master model. The linear interpolation is performed ...
```
Would there be 3 columns for the 3D case?

bwspenc commented on modules/tensor_mechanics/include/materials/ComputeEigenstrainBeamFromVariable.h:
```diff
@@ -0,0 +1,45 @@
+InputParameters validParams<ComputeEigenstrainBeamFromVariable>();
+
+/**
+ * ComputeEigenstrainBeamFromVariable computes an Eigenstrain from displacement and rotational
```
Eigenstrain->eigenstrain

bwspenc commented on modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenstrainBeamBase.h:
```diff
@@ -0,0 +1,55 @@
+   * expansion coefficient.
+   * param thermal_strain    The current total linear thermal strain
+   *                         (\delta L / L)
+   * param instantaneous_cte The current instantaneous coefficient of thermal
```
I don't think there's any need for `instantaneous_cte`. It's there in the solid version of this for computing fracture integrals, but that's not applicable for beams.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeEigenstrainBeamFromVariable.C:
```diff
@@ -0,0 +1,78 @@
+{
+  InputParameters params = validParams<ComputeEigenstrainBeamBase>();
+  params.addClassDescription("Computes an eigenstrain from a set of variables");
+  params.addCoupledVar("disp_eigenstrain",
```
I'd rename these to `displacement_eigenstrain_variables` and `rotational_eigenstrain_variables`.

bwspenc commented on modules/tensor_mechanics/src/transfers/VectorPostprocessorTransfer.C:
```diff
@@ -0,0 +1,268 @@
+      "master_component", 2, "Component along which interpolation occurs in the master model.");
+  params.addParam<unsigned int>(
+      "subapp_component", 1, "Component along which interpolation occurs in the SubApp.");
+  params.addRequiredParam<FileName>(
```
The part about listing the positions in the other 2 components is a little confusing. Could we just input all 3 components, and have the component in the direction of the components we are interpolating along be an offset if it is nonzero?

sveerara replied:
Yes, the offset would probably be less confusing. I will make these changes. Thanks!

bwspenc reviewed (commented):
Looks good in general.  Just a few things I've noticed in my first pass through this.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/181364/) on d967087 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11336/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11336/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d41d2cae2b91b5067209c088dc3601dde0352d4c`

sveerara commented:
@bwspenc, I have addressed your comments on this PR. Thanks!

acasagran commented:
Everything looks good to me.

sveerara commented:
Thanks @acasagran for reviewing the PR.

bwspenc commented on modules/tensor_mechanics/doc/content/documentation/systems/Materials/ComputeEigenstrainBeamFromCSVInterpolator.md:
```diff
@@ -0,0 +1,17 @@
+
+## Description
+
+`ComputeEigenstrainBeamFromCSVInterpolator` computes the displacement and rotational eigenstrains from the corresponding csv files. [CSVInterpolator](/CSVInterpolator.md) userobject first reads the csv files and creates a bilinear interpolation object as a function of position and time for each variable. For each quadrature point and time, the interpolated values of the variable (using the qp poi ...
```
change
from the corresponding csv files. [CSVInterpolator](/CSVInterpolator.md) userobject first
to
from a set of csv files. This object uses a [CSVInterpolator](/CSVInterpolator.md) userobject, which first

bwspenc commented on modules/tensor_mechanics/src/transfers/VectorPostprocessorTransfer.C:
```diff
@@ -0,0 +1,299 @@
+//* This file is part of the MOOSE framework
```
A few thoughts on this transfer:

1. This seems like something that may have general utility in MOOSE beyond tensor_mechanics appliations, and should probably be put in the framework instead. It looks like there already is a MultiAppVectorPostprocessorTransfer in MOOSE that transfers from a VPP in the subapp to a VPP in the master app.
2) The special feature of this transfer is that this transfers to an AuxVariable rather than a VPP. Do you think there are applications of this to problems that might not just have line elements like our problem of interest?
3) I think this doesn't actually work with any VPP in general -- it has to work with VPPs that inherit from SamplerBase. You should probably check for that. Also, those samplers all output a (poorly named) vector called 'id' which gives the position along the line, so if you used that for your coordinate, you would not need `subapp_component`, and that component would not be restricted to the Cartesian axes.
4) There's also an argument for not having the `master_component` restricted to the Cartesian axes. You could have the user supply a vector the gives the direction in the master app for the VPP quantities to be interpolated along.

bwspenc commented on modules/tensor_mechanics/src/userobjects/CSVInterpolator.C:
```diff
@@ -0,0 +1,157 @@
+//* This file is part of the MOOSE framework
```
A couple of thoughts on this:

1. This also seems like a general MOOSE capability, rather than something specific to tensor_mechanics.
2. There is already a PiecewiseBilinear function that does a similar thing. Can you make use of that, maybe adding an option to do the offsetting you are doing here?

sveerara replied:
The data from the 2D/3D simulation would most likely be obtained using a vector post processor. To use the current version of PiecewiseBilinear function, the data in the multiple csv files would have first have to be rewritten into a single file in a matrix form. Also the x and y values have to provided in the input file.

So, we would either have to add a python/shell script that first converts the data from multiple csv files into matrix format or add this alternate data file format to PiecewiseBilinear function.

bwspenc reviewed (changes requested):
As far as general quality, it looks good. I have a couple of comments on architectural aspects of this that need some more discussion.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/189833/) on b8f5db2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11336/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11336/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d41d2cae2b91b5067209c088dc3601dde0352d4c`

sveerara commented:
@bwspenc, I have edited LineValueSampler to work with MultiAppUserObjectTransfer. One of the good things about this transfer is that it can transfer information from vpp or user object in sub app to both nodal/element aux variables in master app. 

But the spatial value function is set by the sub app. So, the information from the sub app along a line segment can only be transferred to a translated version of the line segment in the master app. So, I cannot map from y direction of the 2D model in the sub app to the z direction of the beam in the master model. But this is not a major problem. We can define the orientation of the beam to match that in the 2D model. 

I am interested in only mapping the values from a boundary in the 2D domain to a set of nodes along the same (or translated) line segment in the master app. However, the MultiAppUserObjectTransfer does not set just set the value of nodes located at the same location as the line segment provided in the sub app vector processor but sets the value for all nodes that fall within the 2D mesh of the sub app. In https://github.com/idaholab/moose/blob/devel/framework/src/transfers/MultiAppUserObjectTransfer.C#L193-L219, there is an outer loop on all the nodes of the master app and if the node falls within the bounding box of a sub app (calculated using the dimensions of the 2D model in our case + any shift in subapp position specified while defining the MultiApp), the value of the variable at that node is obtained from that sub app using spatialValue function. So, if the same master node falls within the bounding box of two separate sub apps, the value of that node gets overwritten. This could be a problem if the beams in the master app are not well separated. 

Please let me know if I should continue trying to use MultiAppUserObjectTransfer.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/192838/) on acc3957 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11336/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11336/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 22d92a9ca0e52f347021104568acb7882224ccd2`

sveerara commented:
@bwspenc and @acasagran, all the tests are passing now. Please review this PR when you find time. Thanks!

permcody commented:
Ping @bwspenc

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/203337/) on 2ce4daf wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11336/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11336/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a2f370d09c9355f4349998d80817e81cf0234b4`

sveerara commented:
@bwspenc and @permcody,  I have updated the PR. 

I am assuming that the bounding boxes of the sub applications do not overlap. If a master node/element lies within the bounding box of one of the sub Apps, that node is projected onto the line  provided as input to the LineValueSampler. If the projected point is not within start and end points of the LineValueSampler, the value for that master node is set to infinity and an error is produced. 

However, I am not throwing an error if a given master node/element does not fall within any of the sub App's bounding boxes. If I do this, the test in moose/test/tests/transfers/multi_app_userobject_transfer/3d_1d fails because in this test values are transferred from the sub app to a part of the master mesh.

I am not sure if the test in Rattlesnake is failing because of this PR. It does use MultiApp but apparently there is some error with file `iqspc_1D_out_initial_solve0_cp/0000_mesh.cpr/1/split-1-0.cpr'

permcody commented:
@sveerara - OK good. You might consider adding a button to all users to throw an error or not. Sometimes users might want to only transfer values back onto a portion of the domain (as is done in that test) and other times they might not expect that and would appreciate an error. I'd have to think more about the implications or other advantages/disadvantages to this approach but it might be worth looking into.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/205144/) on 7f7f1df wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11336/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11336/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a2f370d09c9355f4349998d80817e81cf0234b4`

sveerara commented:
@permcody and @bwspenc, I have updated this PR. I have added an option for the user to turn the error generation on/off for cases where each master node has to be associated with a point in the subApp.

bwspenc commented on framework/src/vectorpostprocessors/LineValueSampler.C:
```diff
@@ -27,11 +29,16 @@ validParams<LineValueSampler>()
   Point start_point = getParam<Point>("start_point");
   Point end_point = getParam<Point>("end_point");
 
+  if (_line_vector_norm == 0.0)
```
I'd use a fuzzy comparison here:
```
if MooseUtils::absoluteFuzzyEqual(_line_vector_norm, 0.0)

sveerara replied:
I have made this change. Thanks for catching this error.

bwspenc commented on framework/src/transfers/MultiAppUserObjectTransfer.C:
```diff
@@ -37,14 +39,20 @@ validParams<MultiAppUserObjectTransfer>()
       "The UserObject you want to transfer values from.  Note: This might be a "
       "UserObject from your MultiApp's input file!");
-
+  params.addParam<bool>("injective_master_to_sub_map",
```
I don't feel too strongly about this, but I had to go look up what 'injective' means. If there's another term that might be more widely understood, that could be helpful.  Maybe something like `all_master_nodes_contained_in_sub_app`?

bwspenc replied:
@permcody Any thoughts on the name for this parameter?

permcody replied:
I like the latter parameter name, but I'm ok with either.

sveerara replied:
I have changed the name of this parameter to `all_master_nodes_contained_in_sub_app`. Also, it will now generate an error if master node is found in 2 or more sub apps and not just if the master node is not found in any sub app.

bwspenc commented on test/tests/transfers/multiapp_userobject_transfer/tests:
```diff
@@ -23,4 +23,12 @@
+    input = '3d_1d_master.i'
+    cli_args = 'Transfers/layered_transfer_from_sub_app/injective_master_to_sub_map=true'
+    expect_err = 'MultiAppUserObjectTransfer: Master element with centroid at'
+    prereq = 3d_1d
```
On these new tests, let's add the 'requirement', 'design', and 'issues' parameters. Look at moose/test/tests/postprocessors/num_iterations/tests for an example.

sveerara replied:
I have added these additional parameters to the test files. I have also added documentation for MultiAppUserObjectTransfer and LineValueSampler based on my understanding of the these classes. Please let me know if any changes need to be made to test descriptions or added documentation.

bwspenc commented on modules/combined/test/tests/beam_eigenstrain_transfer/master_uo_transfer.i:
```diff
@@ -0,0 +1,233 @@
+    direction = from_multiapp
+    user_object = axial_str
+    multi_app = sub
+    variable = to_var
```
Where it makes sense, let's use that injective option in these tests.

sveerara replied:
I have made this change.

bwspenc reviewed (changes requested):
Looking really good -- just a few minor things.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/206426/) on eeb127f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11336/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11336/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a2f370d09c9355f4349998d80817e81cf0234b4`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/206471/) on c5afba6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11336/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11336/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a2f370d09c9355f4349998d80817e81cf0234b4`

bwspenc commented on modules/tensor_mechanics/test/tests/beam/eigenstrain/tests:
```diff
@@ -0,0 +1,33 @@
+
+    requirement = "ComputeEigenstrainBeamFromVariable shall correctly transfer eigenstrains from"
+                  " auxvariables into eigenstrain material property."
+    design = "ComputeEigenstrainBeamFromVariable"
```
ComputeEigenstrainBeamFromVariable.md?

bwspenc commented on framework/src/transfers/MultiAppUserObjectTransfer.C:
```diff
@@ -171,6 +179,78 @@ MultiAppUserObjectTransfer::execute()
+                BoundingBox app_box = _multi_app->getBoundingBox(i, _displaced_source_mesh);
+
+                if (app_box.contains_point(*node))
+                  node_found_in_sub_app += 1;
```
++node_found_in_sub_app;

bwspenc commented on framework/src/transfers/MultiAppUserObjectTransfer.C:
```diff
@@ -171,6 +179,78 @@ MultiAppUserObjectTransfer::execute()
+                BoundingBox app_box = _multi_app->getBoundingBox(i, _displaced_source_mesh);
+
+                if (app_box.contains_point(centroid))
+                  elem_found_in_sub_app += 1;
```
++elem_found_in_sub_app;

bwspenc commented on framework/src/vectorpostprocessors/LineValueSampler.C:
```diff
@@ -65,3 +72,38 @@ LineValueSampler::generatePointsAndIDs(const Point & start_point,
+LineValueSampler::getValue(Point p) const
+{
+  if (_values.size() != 1)
+    mooseError("LineValueSampler: When using LineValueSampler with MultiAppUserObjectTransfer, "
```
The `getValue()` method could be called by classes other than `MultiAppUserObjectTransfer`, so make these messages more general:

"When calling getValue() on LineValueSampler, only one variable can be provided as input to LineValueSampler"

Same deal with the error below.

bwspenc reviewed (commented):
Just a few really small things that should just take a minute to fix.

sveerara commented:
@bwspenc, I have updated the error message on the tests. All the tests are passing now.

permcody commented:
Thanks @sveerara - Nice work and documentation!
